File: course_3/module_1/programming_assignment_1/solution_2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

// Job represents a job with weight and length.
type Job struct {
	Weight int
	Length int
}

// WeightLengthRatio calculates the ratio (weight / length).
func (j Job) WeightLengthRatio() float64 {
	if j.Length == 0 {
		panic("Length of job cannot be zero")
	}
	return float64(j.Weight) / float64(j.Length)
}

// Scheduler handles job scheduling to minimize weighted completion time.
type Scheduler struct {
	Jobs []Job
}

// NewScheduler creates a new Scheduler instance.
func NewScheduler(jobs []Job) *Scheduler {
	return &Scheduler{Jobs: jobs}
}

// ScheduleJobs sorts jobs based on the ratio (weight / length) in descending order.
func (s *Scheduler) ScheduleJobs() {
	sort.Slice(s.Jobs, func(i, j int) bool {
		return s.Jobs[i].WeightLengthRatio() > s.Jobs[j].WeightLengthRatio()
	})
}

// CalculateWeightedCompletionTime calculates the total weighted completion time.
func (s *Scheduler) CalculateWeightedCompletionTime() int {
	totalTime := 0
	completionTime := 0

	for _, job := range s.Jobs {
		completionTime += job.Length
		totalTime += job.Weight * completionTime
	}

	return totalTime
}

// ReadJobsFromFile reads jobs from a file and returns a list of Job objects.
func ReadJobsFromFile(filePath string) ([]Job, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	var jobs []Job
	scanner := bufio.NewScanner(file)
	var lineNum int

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		lineNum++

		if lineNum == 1 {
			// First line is the number of jobs, skip it.
			continue
		}

		parts := strings.Fields(line)
		if len(parts) != 2 {
			return nil, fmt.Errorf("invalid line format at line %d: %s", lineNum, line)
		}

		weight, err := strconv.Atoi(parts[0])
		if err != nil {
			return nil, fmt.Errorf("invalid weight at line %d: %s", lineNum, parts[0])
		}

		length, err := strconv.Atoi(parts[1])
		if err != nil {
			return nil, fmt.Errorf("invalid length at line %d: %s", lineNum, parts[1])
		}

		jobs = append(jobs, Job{Weight: weight, Length: length})
	}

	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("error reading file: %w", err)
	}

	if len(jobs) == 0 {
		return nil, fmt.Errorf("file contains no jobs")
	}

	return jobs, nil
}

func main() {
	// Explicitly set the file path.
	filePath := "course_3/module_1/programming_assignment_1/jobs.txt"

	// Read jobs from the specified file.
	jobs, err := ReadJobsFromFile(filePath)
	if err != nil {
		fmt.Printf("Error: %s\n", err)
		os.Exit(1)
	}

	// Initialize the scheduler.
	scheduler := NewScheduler(jobs)

	// Schedule jobs and calculate the weighted completion time.
	scheduler.ScheduleJobs()
	totalWeightedCompletionTime := scheduler.CalculateWeightedCompletionTime()

	// Print the total weighted completion time.
	fmt.Printf("Total Weighted Completion Time: %d\n", totalWeightedCompletionTime)
}

File: course_3/module_1/programming_assignment_1/solution_2.py
class Job:
    """
    Represents a job with a weight and length.
    """
    def __init__(self, weight, length):
        self.weight = weight
        self.length = length

    @property
    def weight_length_ratio(self):
        """
        Calculates the ratio (weight / length) for sorting.
        """
        return self.weight / self.length

    def __repr__(self):
        return f"Job(weight={self.weight}, length={self.length})"


class Scheduler:
    """
    Scheduler to minimize the weighted sum of completion times using the ratio-based greedy algorithm.
    """
    def __init__(self, jobs):
        """
        Initializes the scheduler with a list of jobs.
        :param jobs: List of Job objects.
        """
        self.jobs = jobs

    def schedule_jobs(self):
        """
        Schedules jobs based on the ratio (weight / length) in descending order.
        :return: A list of jobs in the scheduled order.
        """
        self.jobs.sort(
            key=lambda job: job.weight_length_ratio,
            reverse=True
        )
        return self.jobs

    def calculate_weighted_completion_time(self):
        """
        Calculates the weighted sum of completion times for the scheduled jobs.
        :return: The weighted sum of completion times as an integer.
        """
        total_time = 0
        completion_time = 0

        for job in self.jobs:
            completion_time += job.length
            total_time += job.weight * completion_time

        return total_time


# Example usage
if __name__ == "__main__":
    # The file path.
    file_path = "jobs.txt"

    # Reading jobs from the specified file.
    jobs = []
    try:
        with open(file_path, "r") as file:
            lines = file.readlines()
            num_jobs = int(lines[0].strip())
            for line in lines[1:]:
                weight, length = map(int, line.split())
                jobs.append(Job(weight, length))
    except FileNotFoundError:
        print(f"Error: The file '{file_path}' does not exist.")
        exit(1)
    except ValueError as e:
        print(f"Error: Invalid file format. {e}")
        exit(1)

    # Scheduling and calculating weighted completion time.
    scheduler = Scheduler(jobs)
    scheduler.schedule_jobs()
    total_weighted_completion_time = scheduler.calculate_weighted_completion_time()

    print(f"Total Weighted Completion Time: {total_weighted_completion_time}")
